File: trip-kit-booking/src/main/java/com/skedgo/tripkit/booking/viewmodel/BookingViewModel.java
package com.skedgo.tripkit.booking.viewmodel;

import com.skedgo.tripkit.booking.BookingForm;
import com.skedgo.tripkit.booking.LinkFormField;

import io.reactivex.Flowable;

public interface BookingViewModel {
    Flowable<BookingForm> bookingForm();

    Flowable<Param> nextBookingForm();

    Flowable<BookingForm> loadForm(Param param);

    Flowable<Boolean> isDone();

    void observeAuthentication(AuthenticationViewModel authenticationViewModel);

    Flowable<Boolean> performAction(BookingForm bookingForm);

    Flowable<Boolean> performAction(LinkFormField linkFormField);

    Flowable<Boolean> isFetching();

    Param paramFrom(BookingForm form);

}

File: trip-kit-booking/src/main/java/com/skedgo/tripkit/booking/viewmodel/BookingViewModel.kt
package com.skedgo.tripkit.booking.viewmodel

import com.skedgo.tripkit.booking.BookingForm
import com.skedgo.tripkit.booking.LinkFormField
import io.reactivex.Flowable

interface BookingViewModel {
    fun bookingForm(): Flowable<BookingForm>?

    fun nextBookingForm(): Flowable<Param>?

    fun loadForm(param: Param): Flowable<BookingForm>?

    fun isDone(): Flowable<Boolean>?

    fun observeAuthentication(authenticationViewModel: AuthenticationViewModel)

    fun performAction(bookingForm: BookingForm): Flowable<Boolean>?

    fun performAction(linkFormField: LinkFormField): Flowable<Boolean>?

    fun isFetching(): Flowable<Boolean>?

    fun paramFrom(form: BookingForm): Param?
}